index.d.ts

```diff
@@ -1,197 +1,215 @@
+}
 
-  interface IHelmetSetIfFunction {
-    (req: fastify.FastifyRequest<http.IncomingMessage>, res: fastify.FastifyReply<http.ServerResponse>): boolean;
-  }
+export interface IHelmetSetIfFunction {
+  (req: FastifyRequest, res: FastifyReply): boolean;
+}
 
-  interface IHelmetHpkpConfiguration {
-    maxAge: number;
-    sha256s: string[];
-    includeSubdomains?: boolean;
-    reportUri?: string;
-    reportOnly?: boolean;
-    setIf?: IHelmetSetIfFunction;
-  }
+export interface IHelmetHpkpConfiguration {
+  maxAge: number;
+  sha256s: string[];
+  includeSubdomains?: boolean;
+  reportUri?: string;
+  reportOnly?: boolean;
+  setIf?: IHelmetSetIfFunction;
+}
 
-  interface IHelmetHstsConfiguration {
-    maxAge?: number;
-    includeSubdomains?: boolean;
-    preload?: boolean;
-    setIf?: IHelmetSetIfFunction;
-    force?: boolean;
-  }
+export interface IHelmetHstsConfiguration {
+  maxAge?: number;
+  includeSubdomains?: boolean;
+  preload?: boolean;
+  setIf?: IHelmetSetIfFunction;
+  force?: boolean;
+}
 
-  interface IHelmetReferrerPolicyConfiguration {
-    policy?: string;
-  }
+export interface IHelmetReferrerPolicyConfiguration {
+  policy?: string;
+}
 
-  interface IHelmetXssFilterConfiguration {
-    setOnOldIE?: boolean;
-  }
+export interface IHelmetXssFilterConfiguration {
+  setOnOldIE?: boolean;
+}
 
-  interface IHelmetExpectCtConfiguration {
-    enforce?: boolean;
-    maxAge?: number;
-    reportUri?: string;
-  }
+export interface IHelmetExpectCtConfiguration {
+  enforce?: boolean;
+  maxAge?: number;
+  reportUri?: string;
+}
+
+export interface FastifyHelmetOptions<
+  RawServer extends fastify.RawServerBase = fastify.RawServerDefault,
+  RawRequest extends fastify.RawRequestDefaultExpression<RawServer> = fastify.RawRequestDefaultExpression<RawServer>,
+  RawReply extends fastify.RawReplyDefaultExpression<RawServer> = fastify.RawReplyDefaultExpression<RawServer>,
+  > {
+  contentSecurityPolicy?: boolean | IHelmetContentSecurityPolicyConfiguration<RawServer, RawRequest, RawReply>;
+  dnsPrefetchControl?: boolean | IHelmetDnsPrefetchControlConfiguration;
+  expectCt?: boolean | IHelmetExpectCtConfiguration;
+  featurePolicy?: IHelmetFeaturePolicyConfigurationStrict | IHelmetFeaturePolicyConfiguration;
+  frameguard?: boolean | IHelmetFrameguardConfiguration;
+  hidePoweredBy?: boolean | IHelmetHidePoweredByConfiguration;
+  hpkp?: boolean | IHelmetHpkpConfiguration;
+  hsts?: boolean | IHelmetHstsConfiguration;
+  ieNoOpen?: boolean;
+  noCache?: boolean;
+  noSniff?: boolean;
+  permittedCrossDomainPolicies?: boolean | IHelmetPermittedCrossDomainPoliciesConfiguration;
+  referrerPolicy?: boolean | IHelmetReferrerPolicyConfiguration;
+  xssFilter?: boolean | IHelmetXssFilterConfiguration;
 }
+
+// We dont know the server type so keep using fastify.RawServerBase (union of all server types) for FastifyHelmetOptions
+declare const fastifyHelmet: FastifyPlugin<FastifyHelmetOptions<fastify.RawServerBase>>;
```

Comment:
We dont know the server type at the time of authoring the plugin so using `fastify.RawServerBase` (which is a union of all server types) for `FastifyHelmetOptions`.

Users of the plugin should then be able to cast req / res to a specific http type